pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -613,4 +594,111 @@ protected void completeCommittingSegments(String realtimeTableName) {
   }
+
+  protected long getPartitionOffset(KafkaStreamMetadata kafkaStreamMetadata, final String offsetCriteria, int partitionId) {
+    final String topicName = kafkaStreamMetadata.getKafkaTopicName();
+    final String bootstrapHosts = kafkaStreamMetadata.getBootstrapHosts();
+
+    return getPartitionOffset(topicName, bootstrapHosts, offsetCriteria, partitionId);
+  }
+
+  private long getPartitionOffset(final String topicName, final String bootstrapHosts, final String offsetCriteria, int partitionId) {
+    SimpleConsumerWrapper kafkaConsumer = SimpleConsumerWrapper.forPartitionConsumption(new KafkaSimpleConsumerFactoryImpl(),
+        bootstrapHosts, "dummyClientId", topicName, partitionId);
+    final long startOffset;
+    try {
+      startOffset = kafkaConsumer.fetchPartitionOffset(offsetCriteria, KAFKA_PARTITION_OFFSET_FETCH_TIMEOUT_MILLIS);
+    } catch (TimeoutException e) {
+      LOGGER.warn("Timed out when fetching partition offsets for topic {} partition {}", topicName, partitionId);
+      throw new RuntimeException(e);
+    } finally {
+      IOUtils.closeQuietly(kafkaConsumer);
+    }
+    return startOffset;
+  }
+
+  /**
+   * Create a consuming segment for the kafka partitions that are missing one.
+   *
+   * @param realtimeTableName is the name of the realtime table (e.g. "table_REALTIME")
+   * @param nonConsumingPartitions is a list of integers (kafka partitions that do not have a consuming segment)
+   * @param llcSegments is a list of segment names in the ideal state as was observed last.
+   */
+  public void createConsumingSegment(String realtimeTableName, List<Integer> nonConsumingPartitions,
+      List<String> llcSegments, KafkaStreamMetadata kafkaStreamMetadata) {
+    List<LLCSegmentName> segmentNames = new ArrayList<>(llcSegments.size());
+    for (String segmentId: llcSegments) {
+      segmentNames.add(new LLCSegmentName(segmentId));
+    }
+
+    Collections.sort(segmentNames, Collections.reverseOrder());
+
+    Collections.sort(nonConsumingPartitions, Collections.reverseOrder());
+
+    final ZNRecord partitionAssignment = getKafkaPartitionAssignment(realtimeTableName);
+    final int nReplicas = partitionAssignment.getListField("0").size(); // Number of replicas of partition 0
+
+    final int nSegments = segmentNames.size();
+    int curPartition = nonConsumingPartitions.get(0);
+    for (int i = 0; i < nSegments; i++) {
+      final LLCSegmentName segmentName = segmentNames.get(i);
+      final int segmentPartitionId = segmentName.getPartitionId();
+      if (segmentPartitionId > curPartition) {
+        continue;
+      } else if (segmentPartitionId < curPartition) {
+        // We went past a partition that has no LLC segments. We can create one with sequence 1
+        int seqNum = STARTING_SEQUENCE_NULMBER;
+        List<String> instances = partitionAssignment.getListField(Integer.toString(curPartition));
+        long startOffset = getPartitionOffset(kafkaStreamMetadata, "smallest", curPartition);
```

Comment:
This should be max of smallest and previous partition end offset, otherwise we may end up with duplicate data.